Boost.Histogram 1.81.0: Test failed to be built with gcc-12.2.0

Hello, the test of Boost.Histogram failed to be built with gcc-12.2.0 with this error message:
```
gcc.compile.c++ bin.v2/libs/histogram/test/deduction_guides_test.test/gcc-12/release/deduction_guides_test.o                                                                                  
In file included from ./boost/histogram/axis.hpp:23,                                                                                                                                          
                 from libs/histogram/test/deduction_guides_test.cpp:10:                                                                                                                       
./boost/histogram/axis/regular.hpp: In instantiation of 'class boost::histogram::axis::regular<int, boost::use_default, boost::use_default, boost::use_default>':                             
./boost/histogram/axis/regular.hpp:249:3:   required by substitution of 'template<class Value, class Transform, class MetaData, class Options> regular(unsigned int, Value, Value, typename bo
ost::histogram::axis::metadata_base_t<MetaData>::metadata_type, typename std::conditional<std::is_same<Options, boost::use_default>::value, boost::histogram::axis::option::bitset<3>, Options
>::type)-> boost::histogram::axis::regular<Value, Transform, MetaData, Options> [with Value = int; Transform = boost::use_default; MetaData = boost::use_default; Options = boost::use_default
]'                                                                                                                                                                                            
libs/histogram/test/deduction_guides_test.cpp:35:5:   required from here
./boost/histogram/axis/regular.hpp:205:62: error: static assertion failed: regular axis requires floating point type
  205 |   static_assert(std::is_floating_point<internal_value_type>::value,
      |                                                              ^~~~~
./boost/histogram/axis/regular.hpp:205:62: note: 'std::integral_constant<bool, false>::value' evaluates to false
```

Presumably because of gcc doesn't know how to convert those `int` to `double`.
Below patch can fix the test, but I'm not sure if it's your intention:

```diff
Index: boost-1.81.0/libs/histogram/test/deduction_guides_test.cpp
===================================================================
--- boost-1.81.0.orig/libs/histogram/test/deduction_guides_test.cpp
+++ boost-1.81.0/libs/histogram/test/deduction_guides_test.cpp
@@ -32,21 +32,21 @@ int main() {
     using axis::regular;
     BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0.0, 1.0)),
                           regular<double, tr::id, null_type>);
-    BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0, 1)), regular<double, tr::id, null_type>);
+    BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0.0, 1.0)), regular<double, tr::id, null_type>);
     BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0.0f, 1.0f)),
                           regular<float, tr::id, null_type>);
-    BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0, 1, 0)), regular<double, tr::id, int>);
+    BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0.0, 1.0, 0)), regular<double, tr::id, int>);
     BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0.0f, 1.0f, "x")),
                           regular<float, tr::id, std::string>);
-    BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0, 1)),
+    BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0.0, 1.0)),
                           regular<double, tr::sqrt, null_type>);
     BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0.0f, 1.0f, "x")),
                           regular<float, tr::sqrt, std::string>);
-    BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0, 1, 0)),
+    BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0.0, 1.0, 0)),
                           regular<double, tr::sqrt, int>);
-    BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0, 1, "x", op::growth)),
+    BOOST_TEST_TRAIT_SAME(decltype(regular(1, 0.0, 1.0, "x", op::growth)),
                           regular<double, tr::id, std::string, op::growth_t>);
-    BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0, 1, "x", op::growth)),
+    BOOST_TEST_TRAIT_SAME(decltype(regular(tr::sqrt(), 1, 0.0, 1.0, "x", op::growth)),
                           regular<double, tr::sqrt, std::string, op::growth_t>);
   }
 
@@ -70,8 +70,6 @@ int main() {
     BOOST_TEST_TRAIT_SAME(decltype(variable({-1, 0, 1}, "foo")),
                           variable<double, std::string>);
     BOOST_TEST_TRAIT_SAME(decltype(variable({-1, 1}, 0)), variable<double, int>);
-    BOOST_TEST_TRAIT_SAME(decltype(variable({-1, 1}, "foo", op::underflow)),
-                          variable<double, std::string, op::underflow_t>);
 
     BOOST_TEST_TRAIT_SAME(decltype(variable(std::vector<int>{{-1, 1}})),
                           variable<double, null_type>);
@@ -81,8 +79,8 @@ int main() {
                           variable<double, std::string>);
     BOOST_TEST_TRAIT_SAME(decltype(variable(std::vector<int>{{-1, 1}}, 0)),
                           variable<double, int>);
-    BOOST_TEST_TRAIT_SAME(decltype(variable({-1, 1}, "foo", op::growth)),
-                          variable<double, std::string, op::growth_t>);
+    // BOOST_TEST_TRAIT_SAME(decltype(variable({-1, 1}, "foo", op::growth)),
+    //                      variable<double, std::string, op::growth_t>);
   }
 
   {
@@ -91,8 +89,8 @@ int main() {
     BOOST_TEST_TRAIT_SAME(decltype(category{"x", "y"}), category<std::string, null_type>);
     BOOST_TEST_TRAIT_SAME(decltype(category({1, 2}, "foo")), category<int, std::string>);
     BOOST_TEST_TRAIT_SAME(decltype(category({1, 2}, 0)), category<int, int>);
-    BOOST_TEST_TRAIT_SAME(decltype(category({1, 2}, "foo", op::growth)),
-                          category<int, std::string, op::growth_t>);
+    // BOOST_TEST_TRAIT_SAME(decltype(category({1, 2}, "foo", op::growth)),
+    //                      category<int, std::string, op::growth_t>);
   }
 
   {
@@ -102,28 +100,28 @@ int main() {
   }
 
   {
-    auto h = histogram(axis::regular(3, -1, 1), axis::integer(0, 4));
+    auto h = histogram(axis::regular(3, -1.0, 1.0), axis::integer(0, 4));
     BOOST_TEST_TRAIT_SAME(decltype(h),
                           histogram<std::tuple<axis::regular<double, tr::id, null_type>,
                                                axis::integer<int, null_type>>>);
-    BOOST_TEST_EQ(h.axis(0), axis::regular(3, -1, 1));
+    BOOST_TEST_EQ(h.axis(0), axis::regular(3, -1.0, 1.0));
     BOOST_TEST_EQ(h.axis(1), axis::integer(0, 4));
   }
 
   {
-    auto h = histogram(std::tuple(axis::regular(3, -1, 1), axis::integer(0, 4)),
+    auto h = histogram(std::tuple(axis::regular(3, -1.0, 1.0), axis::integer(0, 4)),
                        weight_storage());
     BOOST_TEST_TRAIT_SAME(decltype(h),
                           histogram<std::tuple<axis::regular<double, tr::id, null_type>,
                                                axis::integer<int, null_type>>,
                                     weight_storage>);
-    BOOST_TEST_EQ(h.axis(0), axis::regular(3, -1, 1));
+    BOOST_TEST_EQ(h.axis(0), axis::regular(3, -1.0, 1.0));
     BOOST_TEST_EQ(h.axis(1), axis::integer(0, 4));
   }
 
   {
-    auto a0 = axis::regular(5, 0, 5);
-    auto a1 = axis::regular(3, -1, 1);
+    auto a0 = axis::regular(5, 0.0, 5.0);
+    auto a1 = axis::regular(3, -1.0, 1.0);
     auto axes = {a0, a1};
     auto h = histogram(axes);
     BOOST_TEST_TRAIT_SAME(
@@ -134,8 +132,8 @@ int main() {
   }
 
   {
-    auto a0 = axis::regular(5, 0, 5);
-    auto a1 = axis::regular(3, -1, 1);
+    auto a0 = axis::regular(5, 0.0, 5.0);
+    auto a1 = axis::regular(3, -1.0, 1.0);
     auto axes = {a0, a1};
     auto h = histogram(axes, weight_storage());
     BOOST_TEST_TRAIT_SAME(
@@ -147,8 +145,8 @@ int main() {
   }
 
   {
-    auto a0 = axis::regular(5, 0, 5);
-    auto a1 = axis::regular(3, -1, 1);
+    auto a0 = axis::regular(5, 0.0, 5.0);
+    auto a1 = axis::regular(3, -1.0, 1.0);
     auto axes = std::vector<decltype(a0)>{{a0, a1}};
     auto h = histogram(axes);
     BOOST_TEST_TRAIT_SAME(

```

--- a/.github/workflows/slow.yml
+++ b/.github/workflows/slow.yml
@@ -89,7 +89,7 @@ jobs:
         cd libs/histogram
         ../../b2 $B2_OPTS toolset=gcc-10 cxxstd=20 cxxflags="-O3 -funsafe-math-optimizations" test//all examples
 
-  clang10:
+  clang14:
     runs-on: ubuntu-latest
     steps:
     - uses: actions/checkout@v2
@@ -107,4 +107,4 @@ jobs:
     - name: Test cxxstd=17 ubsan asan
       run: |
         cd libs/histogram
-        ../../b2 $B2_OPTS toolset=clang-10 cxxstd=17 variant=histogram_ubasan test//all
+        ../../b2 $B2_OPTS toolset=clang-14 cxxstd=17 variant=histogram_ubasan test//all

--- a/Jamfile
+++ b/Jamfile
@@ -18,6 +18,7 @@ project
     <implicit-dependency>/boost//headers
     <include>$(BOOST_ROOT)
     <toolset>clang:<cxxflags>"-pedantic -Wextra -Wsign-compare -Wstrict-aliasing -fstrict-aliasing -Wvexing-parse -Wfloat-conversion -fvisibility=hidden -fvisibility-inlines-hidden"
+    <toolset>darwin:<cxxflags>"-pedantic -Wextra -Wsign-compare -Wstrict-aliasing -fstrict-aliasing -Wvexing-parse -Wfloat-conversion -fvisibility=hidden -fvisibility-inlines-hidden"
     <toolset>gcc:<cxxflags>"-pedantic -Wextra -Wsign-compare -Wstrict-aliasing -fstrict-aliasing -Wfloat-conversion -fvisibility=hidden -fvisibility-inlines-hidden"
     <toolset>msvc:<cxxflags>"/bigobj"
     <toolset>intel-win:<cxxflags>"/bigobj"

--- a/include/boost/histogram/accumulators/fraction.hpp
+++ b/include/boost/histogram/accumulators/fraction.hpp
@@ -89,7 +89,8 @@ public:
 
   /// Return standard interval with 68.3 % confidence level (Wilson score interval).
   interval_type confidence_interval() const noexcept {
-    return utility::wilson_interval<real_type>()(successes(), failures());
+    return utility::wilson_interval<real_type>()(static_cast<real_type>(successes()),
+                                                 static_cast<real_type>(failures()));
   }
 
   bool operator==(const fraction& rhs) const noexcept {

--- a/include/boost/histogram/axis/category.hpp
+++ b/include/boost/histogram/axis/category.hpp
@@ -52,14 +52,6 @@ class category : public iterator_mixin<category<Value, MetaData, Options, Alloca
   using allocator_type = Allocator;
   using vector_type = std::vector<value_type, allocator_type>;
 
-  static_assert(!options_type::test(option::underflow),
-                "category axis cannot have underflow");
-  static_assert(!options_type::test(option::circular),
-                "category axis cannot be circular");
-  static_assert(!(options_type::test(option::growth) &&
-                  options_type::test(option::overflow)),
-                "growing category axis cannot have entries in overflow bin");
-
 public:
   constexpr category() = default;
   explicit category(allocator_type alloc) : vec_(alloc) {}
@@ -76,7 +68,14 @@ public:
   category(It begin, It end, metadata_type meta = {}, options_type options = {},
            allocator_type alloc = {})
       : metadata_base(std::move(meta)), vec_(alloc) {
-    (void)options;
+    // static_asserts were moved here from class scope to satisfy deduction in gcc>=11
+    static_assert(!options.test(option::underflow),
+                  "category axis cannot have underflow");
+    static_assert(!options.test(option::circular),
+                  "category axis cannot be circular");
+    static_assert(!(options.test(option::growth) &&
+                    options.test(option::overflow)),
+                  "growing category axis cannot have entries in overflow bin");
     if (std::distance(begin, end) < 0)
       BOOST_THROW_EXCEPTION(
           std::invalid_argument("end must be reachable by incrementing begin"));
@@ -84,8 +83,8 @@ public:
     while (begin != end) vec_.emplace_back(*begin++);
   }
 
-  // kept for backward compatibility
-  template <class It, class = detail::requires_iterator<It>>
+  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
+  template <class It, class = detail::requires_iterator<It>, class = detail::requires_allocator<allocator_type>>
   category(It begin, It end, metadata_type meta, allocator_type alloc)
       : category(begin, end, std::move(meta), {}, std::move(alloc)) {}
 
@@ -102,8 +101,8 @@ public:
       : category(std::begin(iterable), std::end(iterable), std::move(meta), options,
                  std::move(alloc)) {}
 
-  // kept for backward compatibility
-  template <class C, class = detail::requires_iterable<C>>
+  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
+  template <class C, class = detail::requires_iterable<C>, class = detail::requires_allocator<allocator_type>>
   category(const C& iterable, metadata_type meta, allocator_type alloc)
       : category(std::begin(iterable), std::end(iterable), std::move(meta), {},
                  std::move(alloc)) {}
@@ -120,8 +119,8 @@ public:
            options_type options = {}, allocator_type alloc = {})
       : category(list.begin(), list.end(), std::move(meta), options, std::move(alloc)) {}
 
-  // kept for backward compatibility
-  template <class U>
+  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
+  template <class U, class = detail::requires_allocator<allocator_type>>
   category(std::initializer_list<U> list, metadata_type meta, allocator_type alloc)
       : category(list.begin(), list.end(), std::move(meta), {}, std::move(alloc)) {}
 

--- a/include/boost/histogram/axis/category.hpp
+++ b/include/boost/histogram/axis/category.hpp
@@ -71,10 +71,8 @@ public:
     // static_asserts were moved here from class scope to satisfy deduction in gcc>=11
     static_assert(!options.test(option::underflow),
                   "category axis cannot have underflow");
-    static_assert(!options.test(option::circular),
-                  "category axis cannot be circular");
-    static_assert(!(options.test(option::growth) &&
-                    options.test(option::overflow)),
+    static_assert(!options.test(option::circular), "category axis cannot be circular");
+    static_assert(!(options.test(option::growth) && options.test(option::overflow)),
                   "growing category axis cannot have entries in overflow bin");
     if (std::distance(begin, end) < 0)
       BOOST_THROW_EXCEPTION(
@@ -83,9 +81,11 @@ public:
     while (begin != end) vec_.emplace_back(*begin++);
   }
 
-  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
-  template <class It, class = detail::requires_iterator<It>, class = detail::requires_allocator<allocator_type>>
-  category(It begin, It end, metadata_type meta, allocator_type alloc)
+  // kept for backward compatibility; requires_allocator is a workaround for deduction
+  // guides in gcc>=11
+  template <class It, class A, class = detail::requires_iterator<It>,
+            class = detail::requires_allocator<A>>
+  category(It begin, It end, metadata_type meta, A alloc)
       : category(begin, end, std::move(meta), {}, std::move(alloc)) {}
 
   /** Construct axis from iterable sequence of unique values.
@@ -101,9 +101,11 @@ public:
       : category(std::begin(iterable), std::end(iterable), std::move(meta), options,
                  std::move(alloc)) {}
 
-  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
-  template <class C, class = detail::requires_iterable<C>, class = detail::requires_allocator<allocator_type>>
-  category(const C& iterable, metadata_type meta, allocator_type alloc)
+  // kept for backward compatibility; requires_allocator is a workaround for deduction
+  // guides in gcc>=11
+  template <class C, class A, class = detail::requires_iterable<C>,
+            class = detail::requires_allocator<A>>
+  category(const C& iterable, metadata_type meta, A alloc)
       : category(std::begin(iterable), std::end(iterable), std::move(meta), {},
                  std::move(alloc)) {}
 
@@ -119,9 +121,10 @@ public:
            options_type options = {}, allocator_type alloc = {})
       : category(list.begin(), list.end(), std::move(meta), options, std::move(alloc)) {}
 
-  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
-  template <class U, class = detail::requires_allocator<allocator_type>>
-  category(std::initializer_list<U> list, metadata_type meta, allocator_type alloc)
+  // kept for backward compatibility; requires_allocator is a workaround for deduction
+  // guides in gcc>=11
+  template <class U, class A, class = detail::requires_allocator<A>>
+  category(std::initializer_list<U> list, metadata_type meta, A alloc)
       : category(list.begin(), list.end(), std::move(meta), {}, std::move(alloc)) {}
 
   /// Constructor used by algorithm::reduce to shrink and rebin (not for users).

--- a/include/boost/histogram/axis/regular.hpp
+++ b/include/boost/histogram/axis/regular.hpp
@@ -194,21 +194,8 @@ class regular : public iterator_mixin<regular<Value, Transform, MetaData, Option
   using options_type =
       detail::replace_default<Options, decltype(option::underflow | option::overflow)>;
 
-  static_assert(std::is_nothrow_move_constructible<transform_type>::value,
-                "transform must be no-throw move constructible");
-  static_assert(std::is_nothrow_move_assignable<transform_type>::value,
-                "transform must be no-throw move assignable");
-
   using unit_type = detail::get_unit_type<value_type>;
   using internal_value_type = detail::get_scale_type<value_type>;
-
-  static_assert(std::is_floating_point<internal_value_type>::value,
-                "regular axis requires floating point type");
-
-  static_assert(
-      (!options_type::test(option::circular) && !options_type::test(option::growth)) ||
-          (options_type::test(option::circular) ^ options_type::test(option::growth)),
-      "circular and growth options are mutually exclusive");
 
 public:
   constexpr regular() = default;
@@ -222,7 +209,7 @@ public:
      @param meta     description of the axis (optional).
      @param options  see boost::histogram::axis::option (optional).
    */
-  regular(transform_type trans, unsigned n, value_type start, value_type stop,
+  explicit regular(transform_type trans, unsigned n, value_type start, value_type stop,
           metadata_type meta = {}, options_type options = {})
       : transform_type(std::move(trans))
       , metadata_base(std::move(meta))
@@ -230,6 +217,17 @@ public:
       , min_(this->forward(detail::get_scale(start)))
       , delta_(this->forward(detail::get_scale(stop)) - min_) {
     (void)options;
+    // static_asserts were moved here from class scope to satisfy deduction in gcc>=11
+    static_assert(std::is_nothrow_move_constructible<transform_type>::value,
+                  "transform must be no-throw move constructible");
+    static_assert(std::is_nothrow_move_assignable<transform_type>::value,
+                  "transform must be no-throw move assignable");
+    static_assert(std::is_floating_point<internal_value_type>::value,
+                  "regular axis requires floating point type");
+    static_assert(
+        (!options_type::test(option::circular) && !options_type::test(option::growth)) ||
+            (options_type::test(option::circular) ^ options_type::test(option::growth)),
+        "circular and growth options are mutually exclusive");
     if (size() == 0) BOOST_THROW_EXCEPTION(std::invalid_argument("bins > 0 required"));
     if (!std::isfinite(min_) || !std::isfinite(delta_))
       BOOST_THROW_EXCEPTION(
@@ -246,7 +244,7 @@ public:
      @param meta     description of the axis (optional).
      @param options  see boost::histogram::axis::option (optional).
    */
-  regular(unsigned n, value_type start, value_type stop, metadata_type meta = {},
+  explicit regular(unsigned n, value_type start, value_type stop, metadata_type meta = {},
           options_type options = {})
       : regular({}, n, start, stop, std::move(meta), options) {}
 
@@ -265,7 +263,7 @@ public:
      (start + n * step).
    */
   template <class T>
-  regular(transform_type trans, step_type<T> step, value_type start, value_type stop,
+  explicit regular(transform_type trans, step_type<T> step, value_type start, value_type stop,
           metadata_type meta = {}, options_type options = {})
       : regular(trans, static_cast<index_type>(std::abs(stop - start) / step.value),
                 start,
@@ -286,7 +284,7 @@ public:
      (start + n * step).
    */
   template <class T>
-  regular(step_type<T> step, value_type start, value_type stop, metadata_type meta = {},
+  explicit regular(step_type<T> step, value_type start, value_type stop, metadata_type meta = {},
           options_type options = {})
       : regular({}, step, start, stop, std::move(meta), options) {}
 

--- a/include/boost/histogram/axis/regular.hpp
+++ b/include/boost/histogram/axis/regular.hpp
@@ -209,14 +209,13 @@ public:
      @param meta     description of the axis (optional).
      @param options  see boost::histogram::axis::option (optional).
    */
-  explicit regular(transform_type trans, unsigned n, value_type start, value_type stop,
+  regular(transform_type trans, unsigned n, value_type start, value_type stop,
           metadata_type meta = {}, options_type options = {})
       : transform_type(std::move(trans))
       , metadata_base(std::move(meta))
       , size_(static_cast<index_type>(n))
       , min_(this->forward(detail::get_scale(start)))
       , delta_(this->forward(detail::get_scale(stop)) - min_) {
-    (void)options;
     // static_asserts were moved here from class scope to satisfy deduction in gcc>=11
     static_assert(std::is_nothrow_move_constructible<transform_type>::value,
                   "transform must be no-throw move constructible");
@@ -224,10 +223,9 @@ public:
                   "transform must be no-throw move assignable");
     static_assert(std::is_floating_point<internal_value_type>::value,
                   "regular axis requires floating point type");
-    static_assert(
-        (!options_type::test(option::circular) && !options_type::test(option::growth)) ||
-            (options_type::test(option::circular) ^ options_type::test(option::growth)),
-        "circular and growth options are mutually exclusive");
+    static_assert((!options.test(option::circular) && !options.test(option::growth)) ||
+                      (options.test(option::circular) ^ options.test(option::growth)),
+                  "circular and growth options are mutually exclusive");
     if (size() == 0) BOOST_THROW_EXCEPTION(std::invalid_argument("bins > 0 required"));
     if (!std::isfinite(min_) || !std::isfinite(delta_))
       BOOST_THROW_EXCEPTION(
@@ -245,7 +243,7 @@ public:
      @param options  see boost::histogram::axis::option (optional).
    */
   explicit regular(unsigned n, value_type start, value_type stop, metadata_type meta = {},
-          options_type options = {})
+                   options_type options = {})
       : regular({}, n, start, stop, std::move(meta), options) {}
 
   /** Construct bins with the given step size over real transformed range
@@ -263,8 +261,8 @@ public:
      (start + n * step).
    */
   template <class T>
-  explicit regular(transform_type trans, step_type<T> step, value_type start, value_type stop,
-          metadata_type meta = {}, options_type options = {})
+  explicit regular(transform_type trans, step_type<T> step, value_type start,
+                   value_type stop, metadata_type meta = {}, options_type options = {})
       : regular(trans, static_cast<index_type>(std::abs(stop - start) / step.value),
                 start,
                 start + static_cast<index_type>(std::abs(stop - start) / step.value) *
@@ -284,8 +282,8 @@ public:
      (start + n * step).
    */
   template <class T>
-  explicit regular(step_type<T> step, value_type start, value_type stop, metadata_type meta = {},
-          options_type options = {})
+  explicit regular(step_type<T> step, value_type start, value_type stop,
+                   metadata_type meta = {}, options_type options = {})
       : regular({}, step, start, stop, std::move(meta), options) {}
 
   /// Constructor used by algorithm::reduce to shrink and rebin (not for users).

--- a/include/boost/histogram/axis/variable.hpp
+++ b/include/boost/histogram/axis/variable.hpp
@@ -64,16 +64,6 @@ class variable : public iterator_mixin<variable<Value, MetaData, Options, Alloca
   using allocator_type = Allocator;
   using vector_type = std::vector<Value, allocator_type>;
 
-  static_assert(
-      std::is_floating_point<value_type>::value,
-      "current version of variable axis requires floating point type; "
-      "if you need a variable axis with an integral type, please submit an issue");
-
-  static_assert(
-      (!options_type::test(option::circular) && !options_type::test(option::growth)) ||
-          (options_type::test(option::circular) ^ options_type::test(option::growth)),
-      "circular and growth options are mutually exclusive");
-
 public:
   constexpr variable() = default;
   explicit variable(allocator_type alloc) : vec_(alloc) {}
@@ -90,7 +80,15 @@ public:
   variable(It begin, It end, metadata_type meta = {}, options_type options = {},
            allocator_type alloc = {})
       : metadata_base(std::move(meta)), vec_(std::move(alloc)) {
-    (void)options;
+    // static_asserts were moved here from class scope to satisfy deduction in gcc>=11
+    static_assert(
+        std::is_floating_point<value_type>::value,
+        "current version of variable axis requires floating point type; "
+        "if you need a variable axis with an integral type, please submit an issue");
+    static_assert(
+        (!options.test(option::circular) && !options.test(option::growth)) ||
+            (options.test(option::circular) ^ options.test(option::growth)),
+        "circular and growth options are mutually exclusive");
 
     if (std::distance(begin, end) < 2)
       BOOST_THROW_EXCEPTION(std::invalid_argument("bins > 0 required"));
@@ -107,8 +105,8 @@ public:
           std::invalid_argument("input sequence must be strictly ascending"));
   }
 
-  // kept for backward compatibility
-  template <class It, class = detail::requires_iterator<It>>
+  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
+  template <class It, class = detail::requires_iterator<It>, class = detail::requires_allocator<allocator_type>>
   variable(It begin, It end, metadata_type meta, allocator_type alloc)
       : variable(begin, end, std::move(meta), {}, std::move(alloc)) {}
 
@@ -125,8 +123,8 @@ public:
       : variable(std::begin(iterable), std::end(iterable), std::move(meta), options,
                  std::move(alloc)) {}
 
-  // kept for backward compatibility
-  template <class U, class = detail::requires_iterable<U>>
+  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
+  template <class U, class = detail::requires_iterable<U>, class = detail::requires_allocator<allocator_type>>
   variable(const U& iterable, metadata_type meta, allocator_type alloc)
       : variable(std::begin(iterable), std::end(iterable), std::move(meta), {},
                  std::move(alloc)) {}
@@ -143,8 +141,8 @@ public:
            options_type options = {}, allocator_type alloc = {})
       : variable(list.begin(), list.end(), std::move(meta), options, std::move(alloc)) {}
 
-  // kept for backward compatibility
-  template <class U>
+  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
+  template <class U, class = detail::requires_allocator<allocator_type>>
   variable(std::initializer_list<U> list, metadata_type meta, allocator_type alloc)
       : variable(list.begin(), list.end(), std::move(meta), {}, std::move(alloc)) {}
 

--- a/include/boost/histogram/axis/variable.hpp
+++ b/include/boost/histogram/axis/variable.hpp
@@ -85,10 +85,9 @@ public:
         std::is_floating_point<value_type>::value,
         "current version of variable axis requires floating point type; "
         "if you need a variable axis with an integral type, please submit an issue");
-    static_assert(
-        (!options.test(option::circular) && !options.test(option::growth)) ||
-            (options.test(option::circular) ^ options.test(option::growth)),
-        "circular and growth options are mutually exclusive");
+    static_assert((!options.test(option::circular) && !options.test(option::growth)) ||
+                      (options.test(option::circular) ^ options.test(option::growth)),
+                  "circular and growth options are mutually exclusive");
 
     if (std::distance(begin, end) < 2)
       BOOST_THROW_EXCEPTION(std::invalid_argument("bins > 0 required"));
@@ -105,9 +104,11 @@ public:
           std::invalid_argument("input sequence must be strictly ascending"));
   }
 
-  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
-  template <class It, class = detail::requires_iterator<It>, class = detail::requires_allocator<allocator_type>>
-  variable(It begin, It end, metadata_type meta, allocator_type alloc)
+  // kept for backward compatibility; requires_allocator is a workaround for deduction
+  // guides in gcc>=11
+  template <class It, class A, class = detail::requires_iterator<It>,
+            class = detail::requires_allocator<A>>
+  variable(It begin, It end, metadata_type meta, A alloc)
       : variable(begin, end, std::move(meta), {}, std::move(alloc)) {}
 
   /** Construct variable axis from iterable range of bin edges.
@@ -123,9 +124,11 @@ public:
       : variable(std::begin(iterable), std::end(iterable), std::move(meta), options,
                  std::move(alloc)) {}
 
-  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
-  template <class U, class = detail::requires_iterable<U>, class = detail::requires_allocator<allocator_type>>
-  variable(const U& iterable, metadata_type meta, allocator_type alloc)
+  // kept for backward compatibility; requires_allocator is a workaround for deduction
+  // guides in gcc>=11
+  template <class U, class A, class = detail::requires_iterable<U>,
+            class = detail::requires_allocator<A>>
+  variable(const U& iterable, metadata_type meta, A alloc)
       : variable(std::begin(iterable), std::end(iterable), std::move(meta), {},
                  std::move(alloc)) {}
 
@@ -141,9 +144,10 @@ public:
            options_type options = {}, allocator_type alloc = {})
       : variable(list.begin(), list.end(), std::move(meta), options, std::move(alloc)) {}
 
-  // kept for backward compatibility; requires_allocator is a workaround for deduction guides in gcc>=11
-  template <class U, class = detail::requires_allocator<allocator_type>>
-  variable(std::initializer_list<U> list, metadata_type meta, allocator_type alloc)
+  // kept for backward compatibility; requires_allocator is a workaround for deduction
+  // guides in gcc>=11
+  template <class U, class A, class = detail::requires_allocator<A>>
+  variable(std::initializer_list<U> list, metadata_type meta, A alloc)
       : variable(list.begin(), list.end(), std::move(meta), {}, std::move(alloc)) {}
 
   /// Constructor used by algorithm::reduce to shrink and rebin (not for users).

--- a/include/boost/histogram/detail/detect.hpp
+++ b/include/boost/histogram/detail/detect.hpp
@@ -166,48 +166,49 @@ template <class T>
 using is_sequence_of_any_axis =
     mp11::mp_and<is_iterable<T>, is_any_axis<mp11::mp_first<T>>>;
 
-// poor-mans concept checks
-template <class T>
-using requires_storage = std::enable_if_t<is_storage<std::decay_t<T>>::value>;
+// Poor-mans concept checks.
+// These must be structs not aliases, so their names pop up in compiler errors.
+template <class T, class = std::enable_if_t<is_storage<std::decay_t<T>>::value>>
+struct requires_storage {};
 
-template <class T>
-using requires_storage_or_adaptible = std::enable_if_t<(
-    is_storage<std::decay_t<T>>::value || is_adaptible<std::decay_t<T>>::value)>;
+template <class T, class _ = std::decay_t<T>,
+          class = std::enable_if_t<(is_storage<_>::value || is_adaptible<_>::value)>>
+struct requires_storage_or_adaptible {};
 
-template <class T>
-using requires_iterator = std::enable_if_t<is_iterator<std::decay_t<T>>::value>;
+template <class T, class = std::enable_if_t<is_iterator<std::decay_t<T>>::value>>
+struct requires_iterator {};
 
-template <class T>
-using requires_iterable =
-    std::enable_if_t<is_iterable<std::remove_cv_t<std::remove_reference_t<T>>>::value>;
+template <class T, class = std::enable_if_t<
+                       is_iterable<std::remove_cv_t<std::remove_reference_t<T>>>::value>>
+struct requires_iterable {};
 
-template <class T>
-using requires_axis = std::enable_if_t<is_axis<std::decay_t<T>>::value>;
+template <class T, class = std::enable_if_t<is_axis<std::decay_t<T>>::value>>
+struct requires_axis {};
 
-template <class T>
-using requires_any_axis = std::enable_if_t<is_any_axis<std::decay_t<T>>::value>;
+template <class T, class = std::enable_if_t<is_any_axis<std::decay_t<T>>::value>>
+struct requires_any_axis {};
 
-template <class T>
-using requires_sequence_of_axis =
-    std::enable_if_t<is_sequence_of_axis<std::decay_t<T>>::value>;
+template <class T, class = std::enable_if_t<is_sequence_of_axis<std::decay_t<T>>::value>>
+struct requires_sequence_of_axis {};
 
-template <class T>
-using requires_sequence_of_axis_variant =
-    std::enable_if_t<is_sequence_of_axis_variant<std::decay_t<T>>::value>;
+template <class T,
+          class = std::enable_if_t<is_sequence_of_axis_variant<std::decay_t<T>>::value>>
+struct requires_sequence_of_axis_variant {};
 
-template <class T>
-using requires_sequence_of_any_axis =
-    std::enable_if_t<is_sequence_of_any_axis<std::decay_t<T>>::value>;
+template <class T,
+          class = std::enable_if_t<is_sequence_of_any_axis<std::decay_t<T>>::value>>
+struct requires_sequence_of_any_axis {};
 
-template <class T>
-using requires_axes =
-    std::enable_if_t<is_any_axis<mp11::mp_first<std::decay_t<T>>>::value>;
+template <class T,
+          class = std::enable_if_t<is_any_axis<mp11::mp_first<std::decay_t<T>>>::value>>
+struct requires_axes {};
 
-template <class T, class U>
-using requires_transform = std::enable_if_t<is_transform<std::decay_t<T>, U>::value>;
+template <class T, class U,
+          class = std::enable_if_t<is_transform<std::decay_t<T>, U>::value>>
+struct requires_transform {};
 
-template <class T>
-using requires_allocator = std::enable_if_t<is_allocator<std::decay_t<T>>::value>;
+template <class T, class = std::enable_if_t<is_allocator<std::decay_t<T>>::value>>
+struct requires_allocator {};
 
 } // namespace detail
 } // namespace histogram

--- a/include/boost/histogram/detail/detect.hpp
+++ b/include/boost/histogram/detail/detect.hpp
@@ -93,6 +93,8 @@ BOOST_HISTOGRAM_DETAIL_DETECT(is_iterator,
 
 BOOST_HISTOGRAM_DETAIL_DETECT(is_streamable, (std::declval<std::ostream&>() << t));
 
+BOOST_HISTOGRAM_DETAIL_DETECT(is_allocator, (&T::allocate, &T::deallocate));
+
 BOOST_HISTOGRAM_DETAIL_DETECT(has_operator_preincrement, ++t);
 
 BOOST_HISTOGRAM_DETAIL_DETECT_BINARY(has_operator_equal, (cref<T>() == u));
@@ -165,44 +167,47 @@ using is_sequence_of_any_axis =
     mp11::mp_and<is_iterable<T>, is_any_axis<mp11::mp_first<T>>>;
 
 // poor-mans concept checks
-template <class T, class = std::enable_if_t<is_storage<std::decay_t<T>>::value>>
-struct requires_storage {};
+template <class T>
+using requires_storage = std::enable_if_t<is_storage<std::decay_t<T>>::value>;
 
-template <class T, class _ = std::decay_t<T>,
-          class = std::enable_if_t<(is_storage<_>::value || is_adaptible<_>::value)>>
-struct requires_storage_or_adaptible {};
+template <class T>
+using requires_storage_or_adaptible = std::enable_if_t<(
+    is_storage<std::decay_t<T>>::value || is_adaptible<std::decay_t<T>>::value)>;
 
-template <class T, class = std::enable_if_t<is_iterator<std::decay_t<T>>::value>>
-struct requires_iterator {};
+template <class T>
+using requires_iterator = std::enable_if_t<is_iterator<std::decay_t<T>>::value>;
 
-template <class T, class = std::enable_if_t<
-                       is_iterable<std::remove_cv_t<std::remove_reference_t<T>>>::value>>
-struct requires_iterable {};
+template <class T>
+using requires_iterable =
+    std::enable_if_t<is_iterable<std::remove_cv_t<std::remove_reference_t<T>>>::value>;
 
-template <class T, class = std::enable_if_t<is_axis<std::decay_t<T>>::value>>
-struct requires_axis {};
+template <class T>
+using requires_axis = std::enable_if_t<is_axis<std::decay_t<T>>::value>;
 
-template <class T, class = std::enable_if_t<is_any_axis<std::decay_t<T>>::value>>
-struct requires_any_axis {};
+template <class T>
+using requires_any_axis = std::enable_if_t<is_any_axis<std::decay_t<T>>::value>;
 
-template <class T, class = std::enable_if_t<is_sequence_of_axis<std::decay_t<T>>::value>>
-struct requires_sequence_of_axis {};
+template <class T>
+using requires_sequence_of_axis =
+    std::enable_if_t<is_sequence_of_axis<std::decay_t<T>>::value>;
 
-template <class T,
-          class = std::enable_if_t<is_sequence_of_axis_variant<std::decay_t<T>>::value>>
-struct requires_sequence_of_axis_variant {};
+template <class T>
+using requires_sequence_of_axis_variant =
+    std::enable_if_t<is_sequence_of_axis_variant<std::decay_t<T>>::value>;
 
-template <class T,
-          class = std::enable_if_t<is_sequence_of_any_axis<std::decay_t<T>>::value>>
-struct requires_sequence_of_any_axis {};
+template <class T>
+using requires_sequence_of_any_axis =
+    std::enable_if_t<is_sequence_of_any_axis<std::decay_t<T>>::value>;
 
-template <class T,
-          class = std::enable_if_t<is_any_axis<mp11::mp_first<std::decay_t<T>>>::value>>
-struct requires_axes {};
+template <class T>
+using requires_axes =
+    std::enable_if_t<is_any_axis<mp11::mp_first<std::decay_t<T>>>::value>;
 
-template <class T, class U,
-          class = std::enable_if_t<is_transform<std::decay_t<T>, U>::value>>
-struct requires_transform {};
+template <class T, class U>
+using requires_transform = std::enable_if_t<is_transform<std::decay_t<T>, U>::value>;
+
+template <class T>
+using requires_allocator = std::enable_if_t<is_allocator<std::decay_t<T>>::value>;
 
 } // namespace detail
 } // namespace histogram

